File: 05-lamp-posts/solution.js
// example input
// lampPosts(4, [[1,1], [2,1], [3,2], [4,2]])
// => 1

// i will assume we cannot have a lampPost with "0" radius
// one method *could* be to build a return array, covering each number as we move

function lampPosts(length, posts){
  let final = new Array(length)
//   here we have an "unlit" array of the appropriate length
// slowly fill array with posts, skipping positions if the light has already filled it 
// (do we need a way to represent being lit, which is separate from containing a post?)
// at the end, send this array through the non-null counter
// iterate through post array and "light" the final one up
    posts.forEach(lamp=>{
        let position = lamp[0]
        let radius = lamp[1]
        let startingPoint = position-radius
        startingPoint<0 ? startingPoint = 0 : null
        let endingPoint = position+radius
        for(startingPoint; startingPoint<endingPoint; startingPoint++){
            if(final[startingPoint]){
                final[startingPoint]['lit'] += 1
            }
            else {
                final[startingPoint] = {'lit':1}
            }
            // track how many of these are lit more than once
            // this is a little wonky still
        }
    })
    return final
}

function countNonNullElements(arr){
    let count = 0
    arr.forEach(x=> x != null? count+=1 : null)
}

console.log(lampPosts(4, [[1,1], [2,1], [3,2], [4,2]]))

File: 05-lamp-posts/solution.rb
def lampposts(int, arr)
    street = {} # something to keep track of what positions are lit and what points aren't
    int.times do |i|
        street[i+1] = nil
    end

    arr = arr.sort{|a, b| b[1] - a[1]} # sorts all lampposts by radius size (largest first)

    arr.each_with_index do |lamppost, index|
        # check what positions are covered by each lamppost
        position = lamppost[0]
        radius = lamppost[1]

        beginning = position - radius < 1 ? 1 : position - radius # no negatives / 0
        ending = (position + radius) > int ? int : (position + radius) # if it's greater than the length of the street, stop at the end
        (beginning..ending).to_a.each do |covered_position| 
            if !street[covered_position]
                street[covered_position] = index
            end
        end
    end

    # determine how many unique lampposts were used
    used_posts = {}
    street.each do |key, value|
        if !value
            used_posts = {}
            break
        else
            used_posts[value] = true
        end
    end
    puts used_posts.length > 0 ? used_posts.length : -1
    return used_posts.length > 0 ? used_posts.length : -1
end

lampposts(5, [[1,2], [2,1], [3,1], [4,2], [5,2]]) # => 2
# (10, [[1,2], [2,1], [3,1], [4,2], [5,2]]) => -1
